opportunity/README.md

```diff
@@ -73,12 +86,69 @@ resources below.
+
+## Async Opportunity Search
+
+- **Conformance URI:** <https://stapi.example.com/v0.1.0/async-opportunities>
+
+STAPI has an optional conformance class providing support for async opportunity
+searches, to accommodate searches for products that require more time than is
+allowable by the duration of an HTTP request. This support requires persisting
+the search state and matching opportunities outside the context of the initial
+request, so they are available for retrival at a later time by the requestor.
+The Opportunity Search Record is an additional entity defined to model the
+required search state.
+
+### Opportunity Search Record
+
+Returned by an async opportunity search. Can also be retrieved directly.
+
+| Field Name | Type                                                                       | Description |
+|------------| -------------------------------------------------------------------------- | ----------- |
+| id         | string                                                                     | **REQUIRED.** Opportunity search record ID. |
+| productId  | string                                                                     | **REQUIRED.** Product identifier. This should be a reference to the [Product](https://github.com/Element84/stapi-spec/blob/main/product/README.md) being searched. |
+| status     | SearchStatus                                                               | **REQUIRED.** The current search status. |
+| links      | \[[Link Object](#link-object)]                                             | List of link objects to resources and related URLs. There **must** be a `rel=opportunities` link to the corresponding opportunity collection when the search is completed. See [Link Object - rel=opportunities](#rel=opportunities) below. |
+
+#### rel=opportunities
+
+This Link object provides the means of retireving the search results when the
+search is completed. That is, it should include the equivalent of `GET
+/products/{productId}/opportunities/{collectionId}` where `productId` is the
+product being searched and `collectionId` is the ID of the opportunity
+collection containing the results of the search.
+
+#### Async search response
+
+In the event of an async opportunity search, `POST
+/products/{productId}/opportunities` must return a 201 status. The body of the
+response must be the created Opportunity Search Record. The `Location` HTTP
+header of the response should point to the direct link of the Opportunity
+Search Record (`/searches/opportunities/{searchRecordId}`).
+
+### Sync vs Async Opportunity Searching
+
+Servers that advertise support for async searching via the conformance class
+must support async searches for all products. That said, servers can still
+default to a sync response for products that support it, if desired.  Clients
+can request sync vs async operation via the HTTP `Prefer` header.
+
+Possible values for the `Prefer` header are `respond-async` or `wait`, where
```

Comment:
Hmmm interesting. Seems like a tradeoff of client complexity for increased provider implementation flexibility? In practice if a client could choose sync or async wouldn't a client always want sync for speed and implementation simplicity if they could get it? What was the use case that led to proposing this workflow?

I do wonder if any provider implementation would actually implement this endpoint in such a way that it could actually respond dynamically with sync or async? Maybe that capability could be sold as a competitive advantage against other providers?

Reply:
While you represent sync as superior, I think there's value to the client in choosing async. With async the search is saved and can be retrieved at a later time to be rerun. The opportunities are also persisted (for some duration, probably not indefinitely) and could be read multiple times. I could imagine a UI client, for example, leveraging async support to allow multiple searches to be executed in parallel, and use the server-side persistence to enable switching between result sets. Additionally, if one product only supports async while others can support sync, it might be easier as a client to standardize on async to simplify the implementation.

> I do wonder if any provider implementation would actually implement this endpoint in such a way that it could actually respond dynamically with sync or async?

Sync is easy to support server side. But if a server has to support async for one product that requires it, then leveraging that async implementation across all other products would be trivial for any clients would prefer async.

Reply:
Oh, forgot to mention, this idea is lifted from the OGC Processes API spec (https://docs.ogc.org/is/18-062r2/18-062r2.html#toc32:~:text=requirement%20A).-,7.11.2.3,-.%C2%A0%20Execution%20mode), which itself based its use of the Prefer header on the [Prefer Header for HTTP RFC](https://www.rfc-editor.org/rfc/rfc7240.txt). That RFC outlines the `request-async` and `wait` preferences.

Reply:
Yeah, fair enough. For the example:

> UI client leveraging async support to allow multiple searches to be executed in parallel

The UI client could do that for sync requests and cache the state locally. It's just moving the state complexity from one place to another. Neither feels necessarily better.

I don't really feel strongly about it in any case, was just looking to explore use cases, so carry on! I imagine this is the type of thing that the spec would get a lot of feedback back about when implementations get built. Perhaps a discussion topic for the next sprint?